File: project_euler/Problem094/Problem094.java
package project_euler;

public class Problem094 {
    public void solution() {
        int target = 1000000000;
        int[] index = new int[]{2, 1};
        long result = 0L;
        while (true) {
            int auxiliary = (index[0] << 1) + index[1] * 3;
            index[1] = index[0] + (index[1] << 1);
            index[0] = auxiliary;

            int value = getTriangle(index, -1, target);
            if (-1 == value) {
                break;
            }
            result += value;

            value = getTriangle(index, 1, target);
            if (-1 == value) {
                break;
            }
            result += value;
        }
        System.out.println(result);
    }

    private int getTriangle(int[] index, int offset, int target) {
        int perimeter = (index[0] + offset) << 1;
        if (perimeter >= target) {
            return -1;
        }

        double length = (double) ((index[0] << 1) + offset) / 3;
        if (Math.abs(Math.floor(length + 0.5) - length) > 0.00000001) {
            return 0;
        }
        double area = (length + offset) * index[1] / 2;
        if (Math.abs(Math.floor(area + 0.5) - area) > 0.00000001) {
            return 0;
        }
        return perimeter;
    }
}

File: project_euler/Problem094/Problem094.py
import math
from typing import List

class Problem094:
    def solution(self) -> None:
        target, index, result = 1000000000, [2, 1], 0
        while True:
            index[0], index[1] = (index[0] << 1) + index[1] * 3, index[0] + (index[1] << 1)
            value = self.__get_triangle(index, -1, target)
            if -1 == value:
                break
            result += value

            value = self.__get_triangle(index, 1, target)
            if -1 == value:
                break
            result += value
        print(result)

    def __get_triangle(self, index: List[int], offset: int, target: int) -> int:
        perimeter = (index[0] + offset) << 1
        if perimeter >= target:
            return -1

        length = ((index[0] << 1) + offset) / 3
        if abs(math.floor(length + 0.5) - length) > 0.00000001:
            return 0
        area = (length + offset) * index[1] / 2
        if abs(math.floor(area + 0.5) - area) > 0.00000001:
            return 0
        return perimeter
